NDTensors/src/lib/Sectors/src/category_product.jl

```diff
@@ -34,27 +49,189 @@ function Base.show(io::IO, s::CategoryProduct)
 function ×(p1::CategoryProduct, p2::CategoryProduct)
   return CategoryProduct(categories_product(categories(p1), categories(p2)))
 end
 
-categories_product(l1::NamedTuple, l2::NamedTuple) = union_keys(l1, l2)
-
-categories_product(l1::Tuple, l2::Tuple) = (l1..., l2...)
-
+×(a, g::AbstractUnitRange) = ×(to_gradedrange(a), g)
+×(g::AbstractUnitRange, b) = ×(g, to_gradedrange(b))
 ×(nt1::NamedTuple, nt2::NamedTuple) = ×(CategoryProduct(nt1), CategoryProduct(nt2))
 ×(c1::NamedTuple, c2::AbstractCategory) = ×(CategoryProduct(c1), CategoryProduct(c2))
 ×(c1::AbstractCategory, c2::NamedTuple) = ×(CategoryProduct(c1), CategoryProduct(c2))
 
-#
-# Dictionary-like implementation
-#
+function ×(l1::LabelledInteger, l2::LabelledInteger)
+  c3 = label(l1) × label(l2)
+  m3 = unlabel(l1) * unlabel(l2)
+  return labelled(m3, c3)
+end
+
+function ×(g1::AbstractUnitRange, g2::AbstractUnitRange)
+  v = map(
+    ((l1, l2),) -> l1 × l2,
+    Iterators.flatten((Iterators.product(blocklengths(g1), blocklengths(g2)),),),
+  )
+  return gradedrange(v)
+end
+
+# ====================================  Fusion rules  ======================================
+# generic case: fusion returns a GradedAxes, even for fusion with Empty
+function fusion_rule(::NotAbelianStyle, s1::CategoryProduct, s2::CategoryProduct)
+  return to_gradedrange(categories_fusion_rule(categories(s1), categories(s2)))
+end
+
+# Abelian case: fusion returns CategoryProduct
+function fusion_rule(::AbelianStyle, s1::CategoryProduct, s2::CategoryProduct)
+  return categories_fusion_rule(categories(s1), categories(s2))
+end
+
+# Empty case
+function fusion_rule(::AbelianStyle, ::TrivialSector, ::TrivialSector)
+  return CategoryProduct(())
+end
+
+# TrivialSector acts as trivial on any AbstractCategory, not just CategoryProduct
+function fusion_rule(::NotAbelianStyle, ::TrivialSector, c::AbstractCategory)
+  return to_gradedrange(c)
+end
+function fusion_rule(::NotAbelianStyle, c::AbstractCategory, ::TrivialSector)
+  return to_gradedrange(c)
+end
+function fusion_rule(::NotAbelianStyle, ::TrivialSector, c::CategoryProduct)
+  return to_gradedrange(c)
+end
+function fusion_rule(::NotAbelianStyle, c::CategoryProduct, ::TrivialSector)
+  return to_gradedrange(c)
+end
+
+# abelian case: return Category
+fusion_rule(::AbelianStyle, c::AbstractCategory, ::TrivialSector) = c
+fusion_rule(::AbelianStyle, ::TrivialSector, c::AbstractCategory) = c
+fusion_rule(::AbelianStyle, c::CategoryProduct, ::TrivialSector) = c
+fusion_rule(::AbelianStyle, ::TrivialSector, c::CategoryProduct) = c
 
+# ===============================  Ordered implementation  =================================
+CategoryProduct(t::Tuple) = _CategoryProduct(t)
+CategoryProduct(cats::AbstractCategory...) = CategoryProduct(cats)
+
+function SymmetryStyle(T::Type{<:Tuple})
+  return mapreduce(SymmetryStyle, combine_styles, fieldtypes(T); init=AbelianStyle())
+end
```

Comment:
I think it would be better to rename this to `categories_symmetrystyle` to follow the convention that we call functions `categories_*(...)` that act on the storage of a `CategoryProduct`.